raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -2268,6 +2269,20 @@ private Long append(int epoch, List<T> records, boolean isAtomic) {
         );
     }
 
+    private void validateSnapshotId(OffsetAndEpoch snapshotId) {
+        Optional<LogOffsetMetadata> highWatermarkOpt = quorum().highWatermark();
+        if (!highWatermarkOpt.isPresent() || highWatermarkOpt.get().offset <= snapshotId.offset) {
+            throw new KafkaException("Trying to creating snapshot with snapshotId: " + snapshotId + " whose offset is larger than the high-watermark: " +
+                    highWatermarkOpt + ". This may necessarily mean a bug in the caller, since the there should be a minimum " +
+                    "size of records between the latest snapshot and the high-watermark when creating snapshot");
+        }
+        int leaderEpoch = quorum().epoch();
+        if (snapshotId.epoch > leaderEpoch) {
+            throw new KafkaException("Trying to creating snapshot with snapshotId: " + snapshotId + " whose epoch is" +
+                    " larger than the current leader epoch: " + leaderEpoch);
+        }
```

Comment:
From the Jira:
> The end offset and epoch of the snapshot is valid based on the leader epoch cache.

How about also validating against the leader epoch cache? See https://github.com/apache/kafka/blob/trunk/raft/src/main/java/org/apache/kafka/raft/ReplicatedLog.java#L124. This is important because both the snapshot and the leader epoch cache are used to validate offsets. See https://github.com/apache/kafka/blob/trunk/raft/src/main/java/org/apache/kafka/raft/ReplicatedLog.java#L85. The term leader epoch cache comes the variable name `leaderEpochCache` used in `kafka.log.Log`.

Reply:
Thanks, previous I thought the quorum epoch is the leader epoch cache as a mistake~
Updated the PR, one thing I'm not sure about is that:

> The epoch of the snapshot is equal to the quorum epoch.

I think the snapshotId's epoch <= quorum epoch should be fine?

Reply:
Yeah, This is not strictly required for correctness. Oh, I see, the check in 2280 is checking that the `epoch > current epoch`.

I mistakenly read it as `epoch != current epoch`. If we perform this check we are basically saying that the caller of `createSnapshot` needs to catch up to the current quorum epoch before it can generate a snapshot. Yes, I think epoch <= quorum epoch` is fine. Let me think about it and I'll update the Jira.